// es-is/index.mjs
// Generated by CoffeeScript 1.12.5
export {
  default as isFunction
} from './function';

export {
  default as isType
} from './type';

export {
  default as isDefined
} from './defined';

export {
  default as isEmpty
} from './empty';

export {
  default as isEmptyArguments
} from './empty-arguments';

export {
  default as isEmptyArray
} from './empty-array';

export {
  default as isEqual
} from './equal';

export {
  default as isHosted
} from './hosted';

export {
  default as isInstanceOf
} from './instanceof';

export {
  default as isNull
} from './null';

export {
  default as isUndefined
} from './undefined';

export {
  default as isArguments
} from './arguments';

export {
  default as isArray
} from './array';

export {
  default as isArrayLike
} from './array-like';

export {
  default as isBool
} from './bool';

export {
  default as isFalse
} from './false';

export {
  default as isTrue
} from './true';

export {
  default as isDate
} from './date';

export {
  default as isValidDate
} from './valid-date';

export {
  default as isElement
} from './element';

export {
  default as isError
} from './error';

export {
  default as isNumber
} from './number';

export {
  default as isInfinite
} from './infinite';

export {
  default as isDecimal
} from './decimal';

export {
  default as isDivisibleBy
} from './divisible-by';

export {
  default as isInteger
} from './integer';

export {
  default as isMax
} from './max';

export {
  default as isMin
} from './min';

export {
  default as isNaN
} from './nan';

export {
  default as isEven
} from './even';

export {
  default as isOdd
} from './odd';

export {
  default as isGe
} from './ge';

export {
  default as isGt
} from './gt';

export {
  default as isLe
} from './le';

export {
  default as isLt
} from './lt';

export {
  default as isWithin
} from './within';

export {
  default as isObject
} from './object';

export {
  default as isPrimitive
} from './primitive';

export {
  default as isPromise
} from './promise';

export {
  default as isHash
} from './hash';

export {
  default as isRegExp
} from './regexp';

export {
  default as isString
} from './string';

export {
  default as isBase64
} from './base64';

export {
  default as isHex
} from './hex';

export {
  default as isSymbol
} from './symbol';
