// es-raf/dist/es-raf.mjs
var browser = (function() {
  var loadTime, now;
  if ((typeof performance !== "undefined" && performance !== null) && performance.now) {
    now = function() {
      return performance.now();
    };
  } else {
    now = function() {
      return Date.now() - loadTime;
    };
    loadTime = new Date().getTime();
  }
  return now;
})();

var cancelAnimationFrame;
var frameDuration;
var id;
var last;
var queue;
var requestAnimationFrame;

frameDuration = 1000 / 60;

id = 0;

last = 0;

queue = [];

var raf = requestAnimationFrame = function(callback) {
  var next, now_;
  if (queue.length === 0) {
    now_ = browser();
    next = Math.max(0, frameDuration - (now_ - last));
    last = next + now_;
    setTimeout(function() {
      var cp, err, i, len, x;
      cp = queue.slice(0);
      queue.length = 0;
      for (i = 0, len = cp.length; i < len; i++) {
        x = cp[i];
        if (!x.cancelled) {
          try {
            x.callback(last);
          } catch (error) {
            err = error;
            setTimeout(function() {
              throw err;
            }, 0);
          }
        }
      }
    }, Math.round(next));
  }
  queue.push({
    handle: ++id,
    callback: callback,
    cancelled: false
  });
  return id;
};

var caf = cancelAnimationFrame = function(handle) {
  var i, len, x;
  for (i = 0, len = queue.length; i < len; i++) {
    x = queue[i];
    if (x.handle === handle) {
      x.cancelled = true;
    }
  }
};

export { raf, caf };
