// postmortem/lib/utils.js
// Generated by CoffeeScript 1.12.4
var coffeeStackRegex, fs, mapSourcePosition, nodeStackRegex, prettyPrint;

fs = require('fs');

mapSourcePosition = require('./source-map-support').mapSourcePosition;

nodeStackRegex = /\n    at [^(]+ \((.*):(\d+):(\d+)\)/;

coffeeStackRegex = /\((.*)\.coffee:(\d+):(\d+)\)/;

prettyPrint = function(err, options) {
  var caret, data, line, match, position;
  if (options == null) {
    options = {};
  }
  if (options.colorize == null) {
    options.colorize = false;
  }
  match = nodeStackRegex.exec(err.stack);
  if (match == null) {
    match = coffeeStackRegex.exec(err.stack);
  }
  if ((match != null) && fs.existsSync(match[1])) {
    position = mapSourcePosition({
      source: match[1],
      line: match[2],
      column: match[3]
    });
    data = fs.readFileSync(position.source, 'utf8');
    if (line = data.split(/(?:\r\n|\r|\n)/)[position.line - 1]) {
      console.error(position.source + ':' + position.line);
      console.error(line);
      if (options.colorize) {
        caret = '\x1B[31m^\x1B[39m';
      } else {
        caret = '^';
      }
      console.error(((new Array(+position.column)).join(' ')) + caret);
    }
  }
  return console.error(err.stack);
};

module.exports = {
  coffeeStackRegex: coffeeStackRegex,
  nodeStackRegex: nodeStackRegex,
  prettyPrint: prettyPrint
};
